Add Multicell location library and sample

jtguggedal opened the pull request:
The PR adds a multicell location library that offers functionality for retrieving a device's location based on LTE cell measurements obtained using the LTE link controller.

The sample shows how the library can be utilized in an application. Location requests are by default triggered by pressing button 1, and can also be configured to happen automatically at a set interval.

Documentation for both library and sample will be added soon.

Currently location services from HERE and Skyhook are suppored, and more may be added before 1.6.0 release.

rlubos commented on lib/multicell_location/services/here_integration.c:
```diff
@@ -0,0 +1,230 @@
+
+static char body[1024];
+static char neighbors[512];
+static const char tls_certificate[] =
```
Do we have information what certificate is this? Probably worth to mention in the comment.

jtguggedal replied:
Good idea, added metadata

rlubos commented on lib/multicell_location/services/skyhook_integration.c:
```diff
@@ -0,0 +1,282 @@
+		"\"serving\": false"					\
+	"}"
+
+static const char tls_certificate[] =
```
Same, would be good to mention what certificate is this.

rlubos commented on lib/multicell_location/services/location_service.h:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @return 0 on success, or negative errror code on failure.
+ */
+int location_service_generete_request(const struct lte_lc_cells_info *cell_data,
```
s/generete/generate

rlubos commented on samples/nrf9160/multicell_location/src/main.c:
```diff
@@ -0,0 +1,235 @@
+		return;
+	}
+
+	printk("Current cell:\n");
```
Why mixing up logger and printk?

jtguggedal replied:
Waited to convert to upmerge is in and logger now handles floats too, changed now

rlubos reviewed (commented):
Will the documentation follow?

jayteemo reviewed (commented):
`samples/nrf9160/multicell_location/README.rst`  is an empty file

tokangas commented on lib/multicell_location/services/location_service.h:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @param cell_data Pointer to neighbor cell data.
+ * @param buf Buffer for storing the HTTP request.
+ * @param buf_len Lenght of the provided buffer.
```
```suggestion
 * @param buf_len Length of the provided buffer.
```

tokangas commented on samples/nrf9160/multicell_location/prj.conf:
```diff
@@ -0,0 +1,63 @@
+
+# General config
+CONFIG_NEWLIB_LIBC=y
+CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
```
I've lately had trouble printing floats if CONFIG_FPU isn't explicitly enabled in the application configuration. You haven't had this issue?

jtguggedal replied:
Added `CONFIG_FPU` as that will now allow printing of floats using the logger

tokangas commented on samples/nrf9160/multicell_location/prj.conf:
```diff
@@ -0,0 +1,63 @@
+
+# HERE location service configuration
+# By default, HERE API v2 with API key authentication is used.
+# This can be changes to v1 and app code + app ID authentication using Kconfig
```
```suggestion
# This can be changed to v1 and app code + app ID authentication using Kconfig
```

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+		REQUIRED = 2,
+	};
+
+        verify = REQUIRED;
```
Wan't aware of that one, thanks!

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+
+static int send_request(const char *request, size_t request_len)
+{
+        int err, fd, bytes;
```
Messed up my editor settings, hopefully fixed now

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	LOG_DBG("Received %d bytes", bytes);
+	LOG_DBG("HTTP response: \n%s\n\n", recv_buf);
```
Thought I fixed that.. Changed to `LOG_DBG("HTTP response: \n%.*s\n", bytes, recv_buf);`

rlubos replied:
But is this only the matter of the debug log? `location_service_parse_response()` uses `strstr` on the `recv_buf` so we should probably assure NULL termination anyway?

jtguggedal replied:
You're right. Added null termination for the receive buffer

jtguggedal commented on lib/multicell_location/services/here_integration.c:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	if (cell_data->ncells_count == 0) {
+		snprintk(body, sizeof(body), HTTP_REQUEST_BODY_NO_NEIGHBORS,
```
Using snprintk all over now, I think. And added error checks

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+        return err;
+}
+
+int multicell_location_get(const struct lte_lc_cells_info *cell_data,
```
Yep, added NULL checks

jtguggedal commented on include/net/multicell_location.h:
```diff
@@ -0,0 +1,63 @@
+ * @return 0 on success, or negative errror code on failure.
+ */
+int multicell_location_get(const struct lte_lc_cells_info *cell_data,
+                           struct multicell_location *result);
```
That sounds better, changed to `location`

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+	LOG_DBG("Sent %d bytes", offset);
+
+	bytes = recv(fd, recv_buf, sizeof(recv_buf), 0);
+	if (bytes < 0) {
```
That's true. Returning `ENODATA` now in case of 0 bytes received

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+
+        err = send_request(http_request, strlen(http_request));
+        if (err) {
+                LOG_ERR("Failed to send HTTP request");
```
Trur that. Changed to `execute_http_request()`

tokangas replied:
Maybe still change the error trace text as well?

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+	LOG_INF("Provisioning certificate");
+
+	/*  Provision certificate to the modem */
+	err = modem_key_mgmt_write(TLS_SEC_TAG,
+				   MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+				   certificate, strlen(certificate));
```
Nice. In this case I need to use `strlen`, as the compiler doesn't know the size of the buffer that's pointed to and `sizeof` would only give the size of the pointer

tokangas replied:
Ah, you're right, my mistake. :)

jtguggedal commented on lib/multicell_location/services/here_integration.c:
```diff
@@ -0,0 +1,230 @@
+	 * {"location":{"lat":<latitude>,"lng":<longitude>,"accuracy":<accuracy>}}
+	 */
+
+	http_status = strstr(response, "200 OK");
```
Ah, forgot to update Here implementation after I discovered that Skyhook does exactly what you way and omit the "OK" :)

jtguggedal commented:
Thanks for reviewing @rlubos @jayteemo  and @tokangas ! I think I have addressed your comments.

Docs will come soon. Adding DNM until then.

tokangas commented on samples/nrf9160/multicell_location/src/main.c:
```diff
@@ -0,0 +1,235 @@
+	} else {
+		int err;
+
+		err = lte_lc_psm_req(true);
```
I started wondering what happens with neighbor cell measurements when PSM and/or eDRX is enabled. When modem is awaken from sleep using %NCELLMEAS, the currently available cells may have changed if the device is moving. Modem will only scan for cells it already knows, so it might be that it is not able to find any cells.

Modem could be forced to do a TAU by changing some parameter or by sending uplink data. Or we could disable PSM and configure the eDRX period to be short enough. Currently there doesn't seem to be a good way around this.

tokangas commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+		LOG_ERR("No data received");
+		err = -ENODATA;
+		goto clean_up;
+
```
nit: Remove empty line.

simensrostad commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+		LOG_ERR("No certificate was provided by the location service");
+		return -EFAULT;
+	}
+
```
The API states that provisioning depends on LTE not being active. Since we are already depending on the link controller why no add a check here and return error if the link is in fact active?

jtguggedal replied:
Added an assert to the sample to make sure auto connect is not used, can add a check here to later on

simensrostad commented on lib/multicell_location/services/location_service.h:
```diff
@@ -0,0 +1,57 @@
+ *	  location structure.
+ *
+ * @param response Response from location service.
+ * @param result Storage for the result from response parsing.
```
```suggestion
 * @param location Storage for the result from response parsing.
```

simensrostad commented on lib/multicell_location/services/skyhook_integration.c:
```diff
@@ -0,0 +1,329 @@
+#include <zephyr.h>
+#include <stdio.h>
+#include <modem/lte_lc.h>
+#include <modem/modem_info.h>
```
In this case `cJSON` is not a hard dependency but rather added if enabled indirectly with `modem_info`.

jtguggedal replied:
Added cJSON and CJSON_os explicitly

simensrostad commented on lib/multicell_location/services/skyhook_integration.c:
```diff
@@ -0,0 +1,329 @@
+		return -1;
+	}
+
+	location_obj = cJSON_GetObjectItemCaseSensitive(root_obj, "location");
```
Potential future improvement: `here_integration.c` and skyhook_integration.c` share common parsing functionality. This could be abstracted to some common parsing library between the two and possible future integration layers that share similar encoding. For easier test-ability and to save some lines.

jtguggedal replied:
Yes, will probably move over to cJSON later, the current snprintk-based stuff isn't optimal

simensrostad commented on samples/nrf9160/multicell_location/src/main.c:
```diff
@@ -0,0 +1,235 @@
+	} else {
+		int err;
+
+		err = lte_lc_psm_req(true);
```
If `CONFIG_LTE_AUTO_INIT_AND_CONNECT` Is enabled the sample will not request PSM nor eDRX. Also the handler will not be registered that `NCELLMEAS` callbacks depend on. It seems that this sample should probably not support this config. The semaphore that blocks the main thread will also not be given. In the case you decide to remove it the sample should probably have an assert in case its enabled by a user.

jtguggedal replied:
Added assert to ensure auto connect isn't used in the sample and removed the check for auto connect here

simensrostad reviewed (commented):
Overall looks good. Left a few comments to consider.

jtguggedal commented on lib/multicell_location/Kconfig:
```diff
@@ -0,0 +1,120 @@
+
+config MULTICELL_LOCATION_SERVICE_HERE
+	bool "HERE location service"
+	select CJSON_LIB
```
Agreed, I moved the cJSON thing to the library menuconfig

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+#include <net/tls_credentials.h>
+#include <modem/modem_key_mgmt.h>
+#include <net/multicell_location.h>
+#include <cJSON_os.h>
```
Yep, moved

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+		LOG_ERR("Failed to setup TLS sec tag, error: %d", errno);
+		return -errno;
+	}
+
```
Thought about it, but might have to leave it as a future improvement for now

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+
+	/* Setup TLS socket options */
+	err = tls_setup(fd);
+	if (err) {
```
Actually removed printing here because the error details are already printed within `tls_setup`, so it just clogged up the log without adding value

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+		return -EINVAL;
+	}
+
+	if (!cjson_is_init) {
```
Yes, we should add that

jtguggedal commented on lib/multicell_location/multicell_location.c:
```diff
@@ -0,0 +1,230 @@
+				    &exists, &unused);
+	if (err) {
+		LOG_ERR("Failed to check for certificates err %d", err);
+		return err;
```
Left as is for now, could optimize it later

jtguggedal commented:
Updated the PR with documentation for library and sample, and addressed comments by @simensrostad.
I will add one more section to the sample docs with some more detail on what to expect in RRC idle vs RRC connected in terms of timing advance and neighbor cells, but the main parts of the docs are ready for review.

simensrostad reviewed (approved):
LGTM, thanks!

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4665
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

jtguggedal commented:
Updated with a section about LTE cell measurements and what to expect in RRC connected vs RRC idle, and a fix to only include timing advance if it has a valid value
